Framework/Algorithms/src/Bin2DPowderDiffraction.cpp

```diff
@@ -0,0 +1,385 @@
+  const std::string beFileName = getProperty("BinEdgesFile");
+  std::ifstream file(beFileName);
+  std::string line;
+  std::string::size_type n;
+  std::string::size_type sz;
+  std::vector<double> tmp;
+  int dpno = 0;
+  while (getline(file, line)) {
+    n = line.find("dp =");
+    if (n != std::string::npos) {
+      if (!tmp.empty()) {
+        Xbins.push_back(tmp);
+        tmp.clear();
+      }
+      double dp1 = std::stod(line.substr(4), &sz); // 4 is needed to crop 'dp='
+      double dp2 = std::stod(line.substr(sz + 4));
+      if (dpno < 1) {
+        Ybins.push_back(dp1);
+        Ybins.push_back(dp2);
+      } else {
+        Ybins.push_back(dp2);
+      }
+      dpno++;
+    } else if (line.find("#") == std::string::npos) {
+      std::stringstream ss(line);
+      double d;
+      while (ss >> d) {
+        tmp.push_back(d);
+      }
+    }
+  }
+  Xbins.push_back(tmp);
+  g_log.information() << "Number of Ybins: " << Ybins.size() << std::endl;
+  g_log.information() << "Number of Xbins sets: " << Xbins.size() << std::endl;
+}
+//----------------------------------------------------------------------------------------------
+/**
+ * @brief Bin2DPowderDiffraction::UnifyXBins unifies size of the vectors in
+ *Xbins.
+ * Fills the last vector element at the end of the shorter bins.
+ * Required to avoid garbage values in the X values after ws->setHistogram.
+ * returns the maximal size
+ *
+ * @param[in] Xbins --- bins to unify. Will be overwritten.
+ */
+size_t Bin2DPowderDiffraction::UnifyXBins(
+    std::vector<std::vector<double>> &Xbins) const {
+  // get maximal vector size
+  size_t max_size = 0;
+  for (const auto &v : Xbins) {
+    max_size = std::max(v.size(), max_size);
+  }
+  // resize all vectors to maximum size, fill last vector element at the end
+  for (auto &v : Xbins) {
+    if (v.size() < max_size)
+      v.resize(max_size, v.back());
+  }
+  return max_size;
+}
+
+void Bin2DPowderDiffraction::normalizeToBinArea(MatrixWorkspace_sptr outWS) {
+  NumericAxis *verticalAxis = dynamic_cast<NumericAxis *>(outWS->getAxis(1));
+  const std::vector<double> yValues = verticalAxis->getValues();
+  auto nhist = outWS->getNumberHistograms();
+  g_log.debug() << "Number of hists: " << nhist
+                << " Length of YAxis: " << verticalAxis->length() << std::endl;
+
+  for (size_t idx = 0; idx < nhist; ++idx) {
+    double factor = 1.0 / (yValues[idx + 1] - yValues[idx]);
+    // divide by the xBinWidth
+    outWS->convertToFrequencies(idx);
+    auto &freqs = outWS->mutableY(idx);
+    std::transform(freqs.begin(), freqs.end(), freqs.begin(),
+                   std::bind1st(std::multiplies<double>(), factor));
+    auto &errors = outWS->mutableE(idx);
+    std::transform(errors.begin(), errors.end(), errors.begin(),
+                   std::bind1st(std::multiplies<double>(), factor));
+  }
+}
+
+void convertToDSpacing(double wavelength, double theta, double *d, double *dp) {
```

Comment:
You could return this as a [`std::pair<double>`](http://en.cppreference.com/w/cpp/utility/pair) which might be slow. Alternatively, you could split them into two separate functions that are called on different lines (L252 and L259) and output to `const double`. Better still would be storing a function so `sin(theta)` and `log(cos(theta))` are only calculated once per spectrum. Examples of this are in the [diffraction utility](https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/inc/MantidKernel/Diffraction.h) that already exists. I'm not sure if this particular functionality should be added there.